vignettes/task_plans.Rmd

```diff
@@ -0,0 +1,353 @@
+
+
+#### Executing a task plan/makefile
+
+You can use `loop_tasks` with the `task_plan` and the `task_makefile` to run all the steps in all the tasks:
+
+```{r delete1, echo=FALSE}
+unlink('.remake')
+unlink('build')
+```
+
+```{r loop1a}
+loop_tasks(
+  task_plan=task_plan_1,
+  task_makefile='task_plan_1.yml')
+```
+
+The benefits of using `loop_tasks` to run all the remake targets in task_plan_1.yml are:
+
+1. baked-in retries during each execution of `loop_tasks`
+2. quicker restarts if `loop_tasks` gets interrupted partway through (e.g., if your computer runs out of battery or crashes)
+3. the ability to easily run subsets of tasks and/or steps while you're developing the analysis (see below)
+
+To use the baked-in retries effectively, simply ensure that the functions called in your task-steps throw errors (typically with `stop()`) when things go wrong. No need to build retries into the functions themselves because `loop_tasks` will take care of that for you. And if your function isn't the sort that would benefit from retries - e.g., HTTP requests need retries, but a deterministic algorithm should just give up if it breaks - you can set `num_tries=1` in `loop_tasks`.
+
+
+### Forcing rebuilds
+
+#### Forcing rebuild for one task-step
+
+If we try to `scmake` any of the targets in the task makefile without changing any of the inputs, `remake` correctly recognizes that they're up to date already, and the target is not rebuilt. For example, because we ran `loop_tasks` above, `AZ_plot.png` is already `OK` when we try to build it again:
+
+```{r try_rebuild_one}
+scmake('AZ_plot.png', 'task_plan_1.yml')
+```
+
+One way to force the rebuild of a single target is to use `scdel`, like this:
+
+```{r delete_one}
+scdel('AZ_plot.png', 'task_plan_1.yml')
+```
+
+Now when we `scmake` that target again, `AZ_plot.png` gets rebuilt:
+
+```{r rebuild_one}
+scmake('AZ_plot.png', 'task_plan_1.yml')
+```
+
+Alternatively, and even faster to type than the 2-line solution above, we can force a rebuild with the `force=TRUE` argument to `scmake`:
+
+```{r force_one}
+scmake('AZ_plot.png', 'task_plan_1.yml', force=TRUE)
+```
+
+#### Forcing rebuild for all task-steps
+
+Similar to the one-target example, above, if we run the same `loop_tasks` command twice, nothing will get rebuilt the second time. Instead, all items will say `OK`:
+
+```{r try_rebuild_all}
+loop_tasks(
+  task_plan=task_plan_1,
+  task_makefile='task_plan_1.yml')
+```
+
+One way to force a complete rebuild (not generally a good use of time, but sometimes necessary) is to use `force=TRUE` argument to `loop_tasks`. This will automatically use `scdel` with all `target_names` in `task_plan_1.yml`. Now when we run the entire loop, everything gets rebuilt:
```

Comment:
Could you give a quick example of when this might be necessary?

Reply:
It could be necessary when your pipeline doesn't capture all the dependencies (not ideal, but also not uncommon). For example, maybe your first task_step is to pull data from NWIS - you'd want force=TRUE when you were ready to pull  a new round of updated data from NWIS, and force=TRUE would be necessary because nothing in your local files exists to tell you that the NWIS data have been changing online.